plasmapy/atomic/nuclear.py

```diff
@@ -88,95 +97,150 @@ def nuclear_reaction_energy(reaction):
+        total_charge = 0
+        for species in species_list:
+            if _is_electron(species):
+                total_charge -= 1
+            elif _is_positron(species):
+                total_charge += 1
+            elif not _is_neutron(species):
+                total_charge += atomic_number(species)
+        return total_charge
+
+    def _mass_energy(species_list):
+        r"""Finds the total mass energy from a list of species, while
+        taking the masses of the fully ionized isotopes."""
+        total_mass = 0.0*units.kg
+        for species in species_list:
+            if _is_electron(species) or _is_positron(species):
+                total_mass += constants.m_e
+            elif _is_neutron(species):
+                total_mass += constants.m_n
+            else:
+                atomic_numb = atomic_number(species)
+                total_mass += ion_mass(species, Z=atomic_numb)
+        return (total_mass * constants.c**2).to(units.J)
+
+    input_err_msg = ("The inputs to nuclear_reaction_energy must be either a "
+                     "a string representing a nuclear reaction (e.g., "
+                     "'D + T -> He-4 + n') or the keywords 'reactants' and "
+                     "'products' as lists with the nucleons or particles "
+                     "involved in the reaction.")
+
+    if kwargs and not args and len(kwargs) == 2:  # keyword inputs
+
+        try:
+            reactants = _get_species(list(kwargs['reactants']))
+            products = _get_species(list(kwargs['products']))
+        except Exception:
+            raise ValueError(input_err_msg)
+
+    elif args and not kwargs and len(args) == 1:  # reaction string input
+
+        try:
+            LHS_string, RHS_string = re.split('-+>', args[0])
+            LHS_list = re.split(' \+ ', LHS_string)
+            RHS_list = re.split(' \+ ', RHS_string)
+        except TypeError:
+            raise TypeError("The left and right hand sides of the reaction "
+                            "should be separated by '-->'")
+
+        try:
+            reactants = _get_species(LHS_list)
+            products = _get_species(RHS_list)
+        except Exception:
+            raise ValueError(input_err_msg)
 
-    try:
-        LHS, RHS = re.split('-+>', reaction)
-    except Exception:
-        raise ValueError("The left and right hand sides of the reaction "
-                         "should be separated by '-->'")
+    else:
+        raise ValueError(input_err_msg)
 
-    reactants = _get_isotopes_list(LHS)
-    products = _get_isotopes_list(RHS)
+    try:
+        if _nucleon_number(reactants) != _nucleon_number(products):
+            raise ValueError("The number of nucleons is not conserved")
 
-    mass_num_reactants = _mass_number_of_list(reactants)
-    mass_num_products = _mass_number_of_list(products)
+        if _total_charge(reactants) != _total_charge(products):
+            raise ValueError("Total charge is not conserved")
```

Comment:
Add tests to `nuclear_reaction_energy` on charge conservation, including one that would raise this as a `ValueError`